### ISSUE TITLE
Unclear dependency

### ISSUE THREAD

**Issue by ramack** [danielwippermann/resol-vbus-java#9]
Currently I don't understand the dependencies between PacketFieldSpec.getType() and PacketFieldSpec.getUnit() (Unit Time/None seem to overlap getType(), what about Error Status, which seems to be best used as string).

Unit=None, Type=Number could be a percentage for pump speed, but what else and how to distinguish?
Or the Resistance value of the EM - how shall I detect them without parsing the name?

Can you explain that a bit?

**Comment by danielwippermann** [comment]
Both `getType()` and `getUnit()` return the respective information stored in the VSF.

There are four supported `Type` variants that simply describe which formatter to use to convert the binary information into human-readable strings.

The `Unit` variants are defined [in the UNIT blocks of the VSF](http://danielwippermann.github.io/resol-vbus/vbus-specification-file-format-v1.html). They summarize all necessary information about a physical unit (a numeric ID for fast comparison, a machine-readable unit code and a human-readable unit text used for formatting).

The normal combination for a field is to use the "Number" type and have a unit that is appended during formatting. Fields, that don't use the "Number" type tend to use the "None" unit to avoid appending any unit suffix.

Regarding your questions:

- If a field is a pump speed (like the MX related 00_0010_7E11_10_0100_076_1_0) it uses the unit "Percent" and type "Number". The unit "Percent" itself then declares that the string "%" is appended during formatting.
- The resistance value of a EM (for example field 00_7E11_6651_10_0100_000_4_0) uses the unit "Ohms" and type "Number". So instead of parsing the field name to detect that case, you can instead either:
    - call `field.getUnit().getUnitId()` and compare the result against the entry of well-known unit IDs from the documentation linked above (in case of Ohms it should be 4), or
    - call `field.getUnit().getUnitCodeText()` to string-compare it against `"Ohms"`

Could you elaborate on your question regarding the Error Status best being a string?

**Comment by ramack** [comment]
Oh, it was late when I asked that. Thanks for your explanation. I also messed up Unit and UnitFamily...

About the error status:
The field "error mask" is defined as 1 byte in the "VBus Protocol Specification", from the library I get multiple fields and their names have an interpretation of the single bits in the mask. Somehow I messed the names up with the value...
Now I understand:
1. the "error mask" is interpreted by resol-vbus-java
2. the type of each error mask field should be Number
3. unit is none (-1)?
4. the value is 0 or 1 and should best be retreived via PacketFieldValue.getRawValueLong()

Is this understanding correct? and does 1. apply for all "mask" fields?

**Comment by danielwippermann** [comment]
Your understanding of 1-4 is correct. The mask used to simply be a 32-bit integer for the user to decode manually. Upon popular request the single bits of **some** of the masks have been decoded in the VBus specification. But the answer to *and does 1. apply for all "mask" fields?* is definitely no. Just search for "00_0010_1050_10_0100_036_2_0" in [the list of supported packet fields](http://danielwippermann.github.io/resol-vbus/vbus-packet-fields.html). Many (if not most) of the "mask" fields still only are the multi-bit representation.

**Comment by ramack** [comment]
Maybe one more question about units:
For temperatures I am used to °C, but even though it is an SI unit some users might prefer °F. What is transmitted on the VBus for temperature fields. Is it ALWAYS °C, or can the unit be °F and is this fixed for the device or does it even depend on the settings in the solar controller?

**Comment by ramack** [comment]
And yet one more thing about units and types:
- how do I detect an interpreted error mask field e. g. Error__Sensor_line_broken of the MX? It has a Type=Number and unit=None (unitID= -1) but it is something special, namely a boolean status and I'd like to use it as such

**Comment by danielwippermann** [comment]
Oops, missed your previous comment, sorry...

> For temperatures I am used to °C, but even though it is an SI unit some users might prefer °F. What is transmitted on the VBus for temperature fields. Is it ALWAYS °C, or can the unit be °F and is this fixed for the device or does it even depend on the settings in the solar controller?

No, it is not always °C (e.g. [00_0010_4223_10_0100_000_2_0](http://danielwippermann.github.io/resol-vbus/vbus-packet-fields.html#0010_4223_0100), but all newer products use °C / Wh and other metric units for communication, regardless of which unit is configured for the local display.

> how do I detect an interpreted error mask field e. g. Error__Sensor_line_broken of the MX? It has a Type=Number and unit=None (unitID= -1) but it is something special, namely a boolean status and I'd like to use it as such

That is simply a matter of forging a "Boolean" enum with two variant for "Yes" and "No" and assigning / forging this enum ID to the appropriate error mask fields.

**Comment by ramack** [comment]
Ok, but the packet 00_0010_4223_10_0100_000_2_0 is ALWAYS °F, as it is sent by the "BTU" variant of the BS Plus and no possibility to dynamically switch, right?

For the Boolean, that would use the values Yes and No for the enum similar to #7. Good. For those I really see no use of two outputs a numerical "0 1" one **and** a boolean type - do you think the assumption that whenever I get "Yes" or "No" from a getEnumVariant it is a bool? Or would "True" and "False" be more safe with this respect?

**Comment by danielwippermann** [comment]
> Ok, but the packet 00_0010_4223_10_0100_000_2_0 is ALWAYS °F, as it is sent by the "BTU" variant of the BS Plus and no possibility to dynamically switch, right?

I assume that the field 00_0010_4223_10_0100_000_2_0 is always °F based on the same assumption you mentioned. But since the ServiceCenter software supports a conversion option called "Treat °C as °F" there seems to be some ancient controller that transmits °F values although the VBus fields specify the value to be °C. But this only is the case for some old legacy products.

>  do you think the assumption that whenever I get "Yes" or "No" from a getEnumVariant it is a bool? Or would "True" and "False" be more safe with this respect?

It would be safe to assume since those fields have a bit mask (`getBitMask() != 0`) limiting it to a single bit and a two-entry `Enum` leaving no choice for a numerical representation :)

**Comment by ramack** [comment]
Thanks for the ting about temperature I will ignore this for now.

About the mask I don't get it. I don't find getBitMask() do you mean PacketTemplateFieldPart.getMask()? - For me it is not obvious how to use it in this case and it somehow feels wrong to use it via the PacketFieldValue instance .

**Comment by danielwippermann** [comment]
Yeah, you are right, I meant `PacketTemplateFieldPart#getMask()`.

The usage I thought off would be something like this (WARNING: untested code):

```
PacketFieldValue pfv = ...;
PacketFieldSpec pfs = pfv.getPacketFieldSpec();
boolean isBitField = ((pfs.getParts().length == 1) && (pfs.getParts [0].getMask() != 0xFF));
Enum enum_ = pfv.getPacketFieldSpec().getEnum();  // proposed API in different issue
boolean isTwoVariantEnum = (enum_ != null) && (enum_.getEnumVariants().length == 2);
boolean isBooleanLikeEnum = isBitField && isTwoVariantEnum;
```

I assume that we could add that as a helper method on the `PacketFieldValue` class. Perhaps something like `isBooleanLikeEnum()` or so... Whether that enum uses "True/False", "Yes/No", "On/Off", "Active/Inactive", "Ok/Faulty" ... does not matter and can be specialized for the field at hand. But that indicator could be used to decided whether or not to provide a numerical accessor at all.

**Comment by ramack** [comment]
I struggled with my memories about seeing fields which are not in that packet an those things which we discussed in the other issue. But if such things don't happen for the bit values and the masks are never used for something like restricting the data range (I.e. can be /=0xFF and still have multiple bits set) this looks feasible. A helper would be perfect of cause.

If we want to absract the string representation I think we need a method returning a bool don't we?

**Comment by danielwippermann** [comment]
I just added an "boolean-like error" enum in ead5f1ece96010a8c27be28c709b341dab766ed7 and applied it to the error mask fields of the DeltaSol MX.

In addition to that the `isBooleanLikeEnum` helper landed in 5a6d5a6d72c5cd5a62a51555054ebefadff78dcb. I incorporated your feedback and now test if the bit count of the mask is equal to one, just in case that in some distant future multi-bit field masks land in the VSF. Feel free to report any issues with the implementation.

> If we want to absract the string representation I think we need a method returning a bool don't we?

Feel free to provide one. I'm hesitant to write it myself because I'm struggling over how to implement it "the right way". Normally it would just be a wrapper around `getRawValueLong` checking its return value for `!= null` and subsequently `!= 0`. But I don't know whether you have additional functionality / type safety in mind. Because I e.g. don't know how the function should react if it is called on a field that is not a boolean-like enum...

**Comment by ramack** [comment]
In my case the boolean cast is anyhow not so interesting (OpenHab has no general boolean type, only OnOff, OpenClosed and alike, which all fit not really to an Error/Ok case).

Therefore this can be closed in my eyes.
